workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/DialogStack.kt

```diff
@@ -0,0 +1,146 @@
+   * can use [ViewTreeLifecycleOwner][androidx.lifecycle.ViewTreeLifecycleOwner] as
+   * usual.
+   */
+  public fun update(
+    overlays: List<Overlay>,
+    viewEnvironment: ViewEnvironment,
+    beforeShowing: () -> Unit = {}
+  ) {
+    // Any nested back stacks have to provide saved state registries of their
+    // own, but these things share a global namespace. To make that practical
+    // we add uniquing strings to the ViewEnvironment for each dialog,
+    // via withBackStackStateKeyPrefix.
+
+    val overlayEnvironments = overlays.mapIndexed { index, _ ->
+      viewEnvironment.withBackStackStateKeyPrefix("[${index + 1}]")
+    }
+
+    // On each update we build a new list of the running dialogs, both the
+    // existing ones and any new ones. We need this so that we can compare
+    // it with the previous list, and see what dialogs close.
+    val newRunners = mutableListOf<DialogRunner<*>>()
+
+    for ((i, overlay) in overlays.withIndex()) {
+      val overlayEnvironment = overlayEnvironments[i]
+
+      newRunners += if (i < runners.size && runners[i].canShowRendering(overlay)) {
+        // There is already a dialog at this index, and it is compatible
+        // with the new Overlay at that index. Just update it.
+        runners[i].also { it.showRendering(overlay, overlayEnvironment) }
+      } else {
+        // We need a new dialog for this overlay. Time to build it.
+        // We wrap our Dialog instances in DialogRunner to keep them
+        // paired with their current overlay rendering and environment.
+        // It would have been nice to keep those in tags on the Dialog's
+        // decor view, more consistent with what ScreenViewFactory does,
+        // but calling Window.getDecorView has side effects, and things
+        // break if we call it to early. Need to store them somewhere else.
+        overlay.toDialogFactory(overlayEnvironment).let { dialogFactory ->
+          DialogRunner(
+            overlay, overlayEnvironment, context, dialogFactory
+          ).also { newRunner ->
+            // Has the side effect of creating the dialog if necessary.
+            newRunner.showRendering(overlay, overlayEnvironment)
+            // Custom behavior from the container to be fired when we show a new dialog.
+            // The default modal container uses this to flush its body of partially
+            // processed touch events that should have been blocked by the modal.
+            beforeShowing()
+            // Show the dialog. We use the container-provided LifecycleOwner
+            // to host an androidx ViewTreeLifecycleOwner and SavedStateRegistryOwner
+            // for each dialog. These are generally expected these days, and absolutely
+            // required by Compose.
+            newRunner.show(getParentLifecycleOwner())
+          }
+        }
+      }
+    }
+
+    (runners - newRunners.toSet()).forEach { it.dismiss() }
```

Comment:
The IDE prompted me to add the `toSet()` call as an optimization, and I'm very suggestible.